### ISSUE TITLE
[wasm32-wasi] About function pointer casting in build/static_src

### ISSUE THREAD

**Issue by lum1n0us** [Nuitka/Nuitka#3012]
I notice there is a widely coding pattern in *build/static_src/CompiledXType.c*. It will cause [a trap during execution](https://webassembly.github.io/spec/core/exec/instructions.html#xref-syntax-instructions-syntax-instr-control-mathsf-call-indirect-x-y)(Step 18.) when using as a wasm32-wasi target.

Take *CompiledFunctionType.c* as an example. 
- Define `Nuitka_Function_get_name` as `static PyObject *Nuitka_Function_get_name(struct Nuitka_FunctionObject *function)`. With one argument.
- When refer it in `static PyGetSetDef Nuitka_Function_getset[]`, cast it to `getter`(from Include/descrobject.h). It means cast `PyObject*(struct Nuitka_FunctionObject*)` -> `PyObject *(PyObject *, void *)`. 
- Therefore, callers will execute `Nuitka_Function_get_name` in a form likes `get(obj, closure)`(`getset_get()` in *Objects/descrobject.c*).

From caller side, callee is a function with two arguments. On the other side, callee is a function with one argument. It works well in C. But it confuses wasm.  

A call with function pointers will be transferred into `call_indirect` in Wasm, and pack the callee type together. During execution, there is a type check between expected function type(like `PyObject*(void*,void*)`) and actually function type(`PyObject*(void*)`) (all pointers are same types). Eventually, the checker will hit a trap because of different amount of arguments.

---
P.S.

I am thinking about using a different function signature for wasm32-wasi target like:

``` c
#if __wasi__
static PyObject *Nuitka_Function_get_name(struct Nuitka_FunctionObject *function, void *data) {
#else
static PyObject *Nuitka_Function_get_name(struct Nuitka_FunctionObject *function) {
#endif
```

**Comment by kayhayen** [comment]
I am OK with these functions getting modified to specify an unused argument instead. But ideally we also it it in a way, where the cast to ``getter`` and ``setter`` becomes unnecessary, that means we accept PyObject *self and cast it to the function object instead, this can only be good at catching errors. If you were to do that, it will be very acceptable.

**Comment by lum1n0us** [comment]
Something like?

``` c
#if __wasi__
static PyObject *Nuitka_Function_get_name(PyObject *self, void *data) {
#else
static PyObject *Nuitka_Function_get_name(PyObject *self,) {
#endif
  struct Nuitka_FunctionObject *function = (struct Nuitka_FunctionObject *)self;
  // ...
```

**Comment by kayhayen** [comment]
Yes, but without the need for the define, just follow the signature of getter and setter:

```
typedef PyObject *(*getter)(PyObject *, void *);
typedef int (*setter)(PyObject *, PyObject *, void *);
```

Trailing commas wouldn't work btw.

**Comment by lum1n0us** [comment]
- There are extra `(getter)` and `(setter)` in *MetaPathBasedLoaderX.c*
- Similar problems happens on `(PyCFunction)` casting. 

Highly likely there are more (not-strictly-compatible)function pointer casting in *build/static_src*.

**Comment by kayhayen** [comment]
You have made excellent progress already. The ones I saw in the loader right now seem standard and PyCFunction looks like it's just more of the same.

**Comment by kayhayen** [comment]
Thanks a lot, sorry for the delay, I was on holiday, 2 weeks of Corona then, and a conference, and needed time to catch up, but I believe with your PR merged for 2.5, this is now solved.
